Carthage poa npos tests

dobertRowneySr opened the pull request:
implements #4269  (testing plan comment)
I did not use `chai-as-promised` package since it wasn't strictly necessary

dobertRowneySr committed:
feat: enable PoA

prevent new era from occurring, effectively:
- stopping validator set election and update
- stopping era payout from being updated
- keeping the same validator set (PoA)

dobertRowneySr committed:
fix: no tips for authorities

dobertRowneySr committed:
fix: formatting

dobertRowneySr committed:
feat: extra script

in order to save new runtime build to docker

dobertRowneySr committed:
fix: add cleanup

dobertRowneySr committed:
fix: rename and add test

dobertRowneySr committed:
Merge branch 'carthage-poa' into carthage-poa-npos-tests

dobertRowneySr committed:
fix: tests

dobertRowneySr committed:
Merge branch 'carthage-poa' into carthage-poa-npos-tests

dobertRowneySr committed:
fix: generate metadata & docker compose

dobertRowneySr committed:
fix: generate metadata

dobertRowneySr committed:
chore: add carthage file scenario

dobertRowneySr committed:
chore: start stakers payout is zero

dobertRowneySr committed:
add test for constant auth set

dobertRowneySr committed:
fix tests and add session tests

dobertRowneySr committed:
feat: add test on forceEra

dobertRowneySr committed:
feat: add tests

dobertRowneySr committed:
feat: add test for zero tips

dobertRowneySr committed:
fix: added bonding ok

dobertRowneySr committed:
tests: add nominate and validate poa

dobertRowneySr committed:
fix:  validation fixture

dobertRowneySr committed:
fix: revert validate test

dobertRowneySr committed:
feat: add validators payout is zero test

dobertRowneySr committed:
fix: re arrange sets

dobertRowneySr committed:
fix: rename file

dobertRowneySr committed:
fix: add test for authorities payouts

dobertRowneySr committed:
feat: add extra NPoS tests

dobertRowneySr committed:
fix: format and finalize tests

dobertRowneySr committed:
feat: install chai-as-promised

dobertRowneySr committed:
feat: add fixtures

dobertRowneySr committed:
fix: finalize test pt2

dobertRowneySr committed:
fix: add test and fix case

dobertRowneySr committed:
fix: finish scenario

dobertRowneySr committed:
feat: add force era change

dobertRowneySr committed:
fix: renaming

dobertRowneySr committed:
chore: formatting

dobertRowneySr committed:
fix: remove tests

dobertRowneySr committed:
fix: authorities getter

dobertRowneySr committed:
feat: authorities allowed to get tips in any case

dobertRowneySr committed:
fix: rearrange tests

dobertRowneySr committed:
fix: misc errors + ignoreVoid linting option

dobertRowneySr committed:
fix: refactor

dobertRowneySr committed:
fix: miscs

dobertRowneySr committed:
feat: tips to authorities are kept

dobertRowneySr committed:
fix: revert ignoreVoid

dobertRowneySr committed:
fix: miscs

dobertRowneySr committed:
fix: better naming

dobertRowneySr committed:
fix: runtime build

dobertRowneySr committed:
fix: clippy

dobertRowneySr committed:
fix: promise returning void

dobertRowneySr committed:
fix: promise returning void

dobertRowneySr committed:
fix: formatting

dobertRowneySr committed:
fix: formatting

dobertRowneySr committed:
fix: flow not returning promise void and ignore promise void

dobertRowneySr committed:
fix: run prettier

mnaamani commented on runtime/src/tests/handle_fees.rs:
```diff
@@ -3,6 +3,7 @@ use super::{
 };
 use crate::{currency, AccountId, Balances, DealWithFees, NegativeImbalance};
 use frame_support::traits::{Currency, Imbalance, OnUnbalanced};
+use sp_runtime::traits::Zero;
```
clippy says drop this import.

mnaamani commented on tests/network-tests/src/scenarios/carthagePoA.ts:
```diff
@@ -0,0 +1,19 @@
+import bondingSucceedsInPoA from 'src/flows/carthage/bondingSucceedsInPoA'
```
change `src/` in import path, to `../` should be  `../flows/carthage/bondingSucceedsInPoA` 

I'm getting `Error: Cannot find module 'src/flows/carthage/bondingSucceedsInPoA'` when trying to run the scenario.

mnaamani replied:
Same issues in other scenario.

For the flows change `src/` to `../../`

Although this problem doesn't show up when running `yarn build` and `yarn lint`, it's caused by the way we run the actual scenario with `node -r ts-node/register  src/scenario-name.ts` which is a bit annoying.

dobertRowneySr replied:
fixed in 810027b7697f1e024fd708c813d0e187490788e6

mnaamani commented on tests/network-tests/src/flows/carthage/constantAuthorities.ts:
```diff
@@ -0,0 +1,22 @@
+  const nextAuthorities = await api.getNextBabeAuthorities()
+  expect(nextAuthorities).to.be.deep.equal(currentAuthorities)
+
+  // make sure that next Session keys are none
```
the comment seems misleading, don't we mean next session keys are the same as current authority session keys?

mnaamani commented on tests/network-tests/src/scenarios/carthageNPoSSwitch.ts:
```diff
@@ -0,0 +1,12 @@
+
+scenario('Carthage NPoS switch', async ({ job }) => {
+  job('set force era to force new', setForceNewEra)
+  job('current era is none', currentEraIsSomeInPoS)
```
jobs in general will run in parallel. In this scenario it makes more sense for job running setForNewEra flow to be the dependency for the remaining jobs. This can be done with the `after` or `require` methods like so:

```
const forceNewEraJob = job('set force era to force new', setForceNewEra)
// will run after forceNewEraJob only if it is successful.
job('current era is none', currentEraIsSomeInPoS).requires(forceNewEraJob)
.
.
```

dobertRowneySr replied:
I put all the PoA / NPoS assertions in one job

mnaamani reviewed (changes requested):
Left a few change requests..

Made local changes to path to run the carthagePoA scenario to just test first two assertions and getting test for current era being none failing:

```
$ node -r ts-node/register --unhandled-rejections=strict src/scenarios/carthagePoA.ts
  integration-tests:api-factory Connecting to chain, attempt 1.. +0ms
Runtime Version: 11.2.0
  integration-tests:scenario Carthage PoA setup +0ms
  integration-tests:job:authority set is constant Running +0ms
  integration-tests:flow: constant Authorities in PoA started +0ms
  integration-tests:job:current era is none Running +0ms
  integration-tests:flow: validator-set started +0ms
  integration-tests:job:current era is none flow 0 failed: +10ms
AssertionError: Unspecified AssertionError
    at currentEraIsNoneInPoA (/Users/mokhtar/joystream/joystream/tests/network-tests/src/flows/carthage/currentEraIsNoneInPoA.ts:14:9)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async /Users/mokhtar/joystream/joystream/tests/network-tests/src/Job.ts:103:11 {
  showDiff: false,
  actual: null,
  expected: undefined
}
  integration-tests:job:current era is none [Failed] +7ms
  integration-tests:job:authority set is constant [Succeeded] +22ms
Job Results:
authority set is constant: Succeeded
current era is none: Failed
Error: Scenario Failed
```

Are tests passing for you?

Also its important to note that the tests are being run with the "TESTING" runtime profile, so no calls are actually filtered in the scenario.

dobertRowneySr committed:
fix: import error and add test scenario to CI checks

mnaamani commented:
It seems even when setting the staking pallet config Forcing to ForceNone, `try_trigger_new_era()` in `staking::new_session()` is being invoked (by session pallet?)

https://github.com/Joystream/substrate/blob/joystream-carthage/frame/staking/src/pallet/impls.rs#L292

This is why testing for current_ers isNone is failing?

dobertRowneySr committed:
fix: bonding amount

dobertRowneySr committed:
fix: scenarios

dobertRowneySr committed:
fix: next active era is none

dobertRowneySr committed:
fix: put together all npos test in one job

dobertRowneySr committed:
fix: almost all poa test succeeding

dobertRowneySr committed:
fix: add test for zero era reward

dobertRowneySr committed:
fix: sudo call works

dobertRowneySr committed:
fix: add test for era reward points

dobertRowneySr committed:
feat: set Era payouts to zero

dobertRowneySr committed:
fix: era payout zero in PoA

dobertRowneySr committed:
fix: adjusted runtime & tests

dobertRowneySr committed:
fix: tests cleanup & format

dobertRowneySr committed:
chore: test cleanup & formatting

dobertRowneySr committed:
chore: test cleanup & formatting

dobertRowneySr committed:
chore: refactor all PoA assertion in one job

dobertRowneySr committed:
chore: refactor all PoA assertion in one job

dobertRowneySr committed:
chore: refactor all PoA assertion in one job

dobertRowneySr committed:
chore: refactor all PoA assertion in one job

dobertRowneySr committed:
chore: refactor all PoA assertion in one job

dobertRowneySr committed:
chore: refactor all PoA assertion in one job

dobertRowneySr committed:
fix: test name

dobertRowneySr committed:
fix: no rewards for initial era

dobertRowneySr committed:
fix: no rewards for initial era

dobertRowneySr committed:
fix: no rewards for initial era

dobertRowneySr committed:
fix: CI errors

dobertRowneySr committed:
fix: CI errors

dobertRowneySr committed:
fix: CI errors

dobertRowneySr committed:
fix: formatting

dobertRowneySr committed:
fix: ci checks

dobertRowneySr committed:
Merge branch 'carthage' into carthage-poa-npos-tests

dobertRowneySr committed:
fix: CI checks

dobertRowneySr committed:
fix: CI checks

dobertRowneySr committed:
fix: CI checks

dobertRowneySr committed:
fix: remove comments

mnaamani commented on tests/network-tests/src/flows/carthage/carthagePoAAssertions.ts:
```diff
@@ -0,0 +1,77 @@
+  debug('started')
+  api.enableDebugTxLogs()
+
+  const bondAmount = new BN(1000000000000000)
```
Perhaps instead of this constant we should take into account the min_validator_bond and min_nominator_bond values, so the test doesn't break if we change those values.
`api.query.staking.minValidatorBond` and `api.query.minNominatorBond` I think.

mnaamani commented on tests/network-tests/src/flows/carthage/carthagePoAAssertions.ts:
```diff
@@ -0,0 +1,77 @@
+  const nextAuthorities = await api.getNextBabeAuthorities()
+  expect(nextAuthorities).to.be.deep.equal(currentAuthorities)
+
+  // 1.b. Next session keys are none
```
The comment doesn't match what is being asserted.
Do we mean, "next authorities are equal to current session authorities" ?

dobertRowneySr replied:
Replaced with : "next session keys (queued keys) are the same as the current sesssion"

mnaamani commented on tests/network-tests/src/flows/carthage/carthagePoAAssertions.ts:
```diff
@@ -0,0 +1,77 @@
+    assert(nextSessionIndex.isNone)
+  }
+
+  // 3. Bonding succeds
```
Of course we want to be testing that bonding works in frozen stage, where balance transfers are not allowed.
Now the testing runtime doesn't filter any calls so this works. Perhaps we want to setup initial balances for the validator and nominator accounts and run the test in the production runtime?

mnaamani commented on tests/network-tests/src/Api.ts:
```diff
@@ -311,6 +315,22 @@ export class Api {
   }
 
+  public async getBabeAuthorities(): Promise<string[]> {
+    return await this.api.query.babe.authorities().then((key) => key.map(([k]) => k.toString()))
```
```suggestion
    return (await this.api.query.babe.authorities()).map(([k]) => k.toString())
```

mnaamani commented on tests/network-tests/src/flows/carthage/carthagePoAAssertions.ts:
```diff
@@ -0,0 +1,77 @@
+  // 5. Nominating succeds:
+  const nominatorCandidatingSucceedsFixture = new NominatingSucceedsFixture(api, [validatorAccount], nominatorAccount)
+  const nominationFixture = new FixtureRunner(nominatorCandidatingSucceedsFixture)
+  await nominationFixture.run()
```
We should also check that validation succeeds, not just bonding

dobertRowneySr replied:
https://crates.parity.io/pallet_staking/pallet/enum.Call.html#variant.validate
Haven't I checked that at assertion 4?

mnaamani commented on tests/network-tests/src/flows/carthage/switchToNPoS.ts:
```diff
@@ -0,0 +1,69 @@
+  // And wait until new era happened: about 10 minutes
+  let forceEraStatus = await api.getForceEra()
+  while (forceEraStatus.isForceNew) {
+    sleep(sleepTimeSeconds * 1000)
```
```suggestion
    await sleep(sleepTimeSeconds * 1000)
```

mnaamani commented on tests/network-tests/src/scenarios/carthageNPoSSwitch.ts:
```diff
@@ -0,0 +1,10 @@
+import switchToNPoS from '../flows/carthage/switchToNPoS'
+
+scenario('Carthage NPoS switch', async ({ job }) => {
+  job('switch to npos checks', switchToNPoS)
```
shouldn't this scenario setup the validators and nominators with running the carthagePoAAssertions before doing the switch? Or will there be an issue with block production after switch if we don't run a second validator?

mnaamani reviewed (changes requested):
Good progress so far. I think the assertions are correct, but in the switch to NPoS I think we want to the test to be a bit more comprehensive.

mnaamani commented:
I just had an observation, this happens both on carthage branch and the current PoA PR, that the initial validators in simple dev chain setup "Alice", although they are correctly producing and finalizing blocks, and they appear in the validators list on the overview page, in the accounts page it doesn't show correctly and we must run validate action again to show up in accounts as validator:

<img width="671" alt="Screen Shot 2022-10-06 at 10 35 32 AM" src="https://user-images.githubusercontent.com/1621012/194231599-fa5de11d-7aa1-499a-bc77-4a20857da412.png">

<img width="723" alt="Screen Shot 2022-10-06 at 10 34 52 AM" src="https://user-images.githubusercontent.com/1621012/194231657-9e0e9a89-bf9a-4bba-af3c-d8fa635ebf4d.png">

<img width="1247" alt="Screen Shot 2022-10-06 at 10 33 10 AM" src="https://user-images.githubusercontent.com/1621012/194231711-1821403c-9108-497e-b954-364896976d20.png">

<img width="864" alt="Screen Shot 2022-10-06 at 10 36 58 AM" src="https://user-images.githubusercontent.com/1621012/194231769-0bc16755-ac4a-442f-b96b-61b5d98090c8.png">

But it looks like it was only to set additional validator preferences:

<img width="679" alt="Screen Shot 2022-10-06 at 10 40 06 AM" src="https://user-images.githubusercontent.com/1621012/194232078-aeb70c95-ec27-4dc9-8c6a-4ab53ddf7458.png">

Although it does look a bit confusing.. cc: @bwhm

Commit:
Update tests/network-tests/src/Api.ts

Co-authored-by: Mokhtar Naamani <mokhtar.naamani@gmail.com>

dobertRowneySr committed:
fix: bond amounts

Commit:
Update tests/network-tests/src/flows/carthage/switchToNPoS.ts

Co-authored-by: Mokhtar Naamani <mokhtar.naamani@gmail.com>

dobertRowneySr committed:
fix: comment

dobertRowneySr committed:
fix: rearrange jobs

dobertRowneySr committed:
fix: errors

Lezek123 commented on runtime/src/lib.rs:
```diff
@@ -537,6 +538,28 @@ pallet_staking_reward_curve::build! {
+        total_issuance: Balance,
+        era_duration_millis: u64,
+    ) -> (Balance, Balance) {
+        let era_index = pallet_staking::Pallet::<Runtime>::current_era().unwrap_or_else(Zero::zero);
```
Based on those definitions of `ActiveEra` and `CurrentEra`:
```rust
	/// The current era index.
	///
	/// This is the latest planned era, depending on how the Session pallet queues the validator
	/// set, it might be active or not.
	#[pallet::storage]
	#[pallet::getter(fn current_era)]
	pub type CurrentEra<T> = StorageValue<_, EraIndex>;

	/// The active era information, it holds index and start.
	///
	/// The active era is the era being currently rewarded. Validator set of this era must be
	/// equal to [`SessionInterface::validators`].
	#[pallet::storage]
	#[pallet::getter(fn active_era)]
	pub type ActiveEra<T> = StorageValue<_, ActiveEraInfo>;
```
And the fact that `end_era` (which tiggers `era_payout`) takes `active_era` as argument, it seems like you should actually query `active_era` here.

I tested this locally and `current_era` changes before `active_era`, causing validators to still recieve rewards for era `0`.

dobertRowneySr replied:
used active era instead

Lezek123 commented on tests/network-tests/.eslintrc.js:
```diff
@@ -8,5 +8,6 @@ module.exports = {
     'new-cap': 'off',
     // Disabled because of the false positive bug: https://github.com/eslint/eslint/issues/11899
     'require-atomic-updates': 'off',
+    '@typescript-eslint/no-floating-promises': 'off',
```
I think that's a good rule to have enabled. If there really needs to have a "floating" promise somewhere, it can be marked with `void`

Lezek123 commented on tests/network-tests/src/fixtures/staking/ValidatingSucceedsFixture.ts:
```diff
@@ -0,0 +1,21 @@
+import { BaseFixture } from '../../Fixture'
+
+export class ValidatingSucceedsFixture extends BaseFixture {
+  protected preferences: any
```
Please try to avoid using `any` type. You can use `createType` utility to construct `ValidatorPrefs` in a typesafe way.

Lezek123 commented on tests/network-tests/src/flows/carthage/carthagePoAAssertions.ts:
```diff
@@ -0,0 +1,78 @@
+  if (activeEra.isSome) {
+    const { index } = activeEra.unwrap()
+    const nextEraIndex = index.addn(1)
+    const nextSessionIndex = await api.getErasStartSessionIndex(nextEraIndex as u32)
```
`nextSessionIndex` is a little confusing name. Perhaps `nextEraStartSessionIndex`?

Lezek123 commented on tests/network-tests/src/flows/carthage/carthagePoAAssertions.ts:
```diff
@@ -0,0 +1,78 @@
+  // 1.b. Queued keys (for next session) and current session keys are the same
+  const sessionAuthorities = await api.getSessionAuthorities()
+  const queuedKeys = await api.getQueuedKeys()
+  expect(queuedKeys).to.be.deep.equal(sessionAuthorities)
```
I think similar checks should also be done **after** some other accounts have already called bond&validate / bond&nominate. So for example:
- get current authorities / sessions authorities _(maybe also imOnline.keys?)_
- use some new accounts to bond & validate/nominate (new validators should have higher stakes than the current ones)
- wait for at least `era_duration` blocks
- make sure the authorities stay unchanged, queuedKeys are equal to current authorities and nextAuthorities are also equal to current authorities

dobertRowneySr replied:
I used the same runtime computation for the `era_duration` locks

Lezek123 commented on tests/network-tests/src/flows/carthage/switchToNPoS.ts:
```diff
@@ -0,0 +1,69 @@
+  while (forceEraStatus.isForceNew) {
+    await sleep(sleepTimeSeconds * 1000)
+    forceEraStatus = await api.getForceEra()
+  }
```
Based on my observations during testing what happens once you call `forceNewEra` is:
- first `staking.forceEra` is set to `ForceNew`
- once the current session ends, `staking.currentEra` is set to `1`, but `activeEra.index` remains `0` (era `1` is just "planned" at this point)
- only after another full session the rewards for era `0` are paid and `activeEra.index` changes to `1`

dobertRowneySr replied:
I used active era as a reference

Lezek123 commented on tests/network-tests/src/flows/carthage/switchToNPoS.ts:
```diff
@@ -0,0 +1,69 @@
+  // 1. Nex Era Starting session index is Some
+  const { index } = (await api.getActiveEra()).unwrap()
+  const nextEraStartingSessionIndex = await api.getErasStartSessionIndex(index.addn(1) as u32)
+  assert(nextEraStartingSessionIndex.isSome, 'ext era starting session index is not some')
```
```suggestion
assert(nextEraStartingSessionIndex.isSome, 'Next era starting session index is not set')
```

Lezek123 commented on tests/network-tests/src/flows/carthage/switchToNPoS.ts:
```diff
@@ -0,0 +1,69 @@
+  const nextEraStartingSessionIndex = await api.getErasStartSessionIndex(index.addn(1) as u32)
+  assert(nextEraStartingSessionIndex.isSome, 'ext era starting session index is not some')
+
+  // 2. Check that genesis authorities claim for era 0 is 0
```
So based on my comment about `forceNewEra`: the rewards for era `0` are not yet set at this point, because `activeEra.index` is still `0`. You should wait until `activeEra.index` is `1` to properly test this.

dobertRowneySr replied:
I have updated the waiting condition to check for active era index > 0

Lezek123 commented on tests/network-tests/src/flows/carthage/switchToNPoS.ts:
```diff
@@ -0,0 +1,69 @@
+  const authorities = await api.getSessionAuthorities()
+  const authoritiesStash = (await Promise.all(authorities.map((account) => api.getBonded(account)))).map((x) =>
+    x.unwrap().toString()
+  )
```
- `await api.getSessionAuthorities()` already returns the stash account
- `staking.bonded` is a map from stash account to controller account, not the other way around

So I think this should be just:
```suggestion
  const authoritiesStash = await api.getSessionAuthorities()
```

Lezek123 commented on tests/network-tests/src/flows/carthage/switchToNPoS.ts:
```diff
@@ -0,0 +1,69 @@
+
+  // 3. Election rounds have happened
+  const electionRounds = await api.getElectionRounds()
+  assert.isAbove(electionRounds.toNumber(), 0)
```
I feel like this test should have more steps, like checking whether the first era rewards work as expected etc.

Lezek123 commented on .github/workflows/run-network-tests.yml:
```diff
@@ -149,7 +149,7 @@ jobs:
     strategy:
       matrix:
-        scenario: ['full', 'setupNewChain', 'setupNewChainMultiStorage', 'bonding']
+        scenario: ['full', 'setupNewChain', 'setupNewChainMultiStorage', 'bonding', 'carthagePoA', 'carthageNPoSSwitch']
```
`carthagePoA` scenario doesn't exist

Commit:
Update tests/network-tests/src/flows/carthage/switchToNPoS.ts

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
fix: era payout

dobertRowneySr committed:
fix: mark eslint with void

dobertRowneySr committed:
fix: validator preferences now typesafe

dobertRowneySr committed:
fix: rename sessionIndex to nextEraStartingIndex

Commit:
Update tests/network-tests/src/flows/carthage/switchToNPoS.ts

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
fix: use active era as reference

dobertRowneySr committed:
chore: remove scenario from CI checks

dobertRowneySr committed:
fix: rearrange tests

dobertRowneySr committed:
fix: add extra test assertions

dobertRowneySr committed:
fix: cleanup

dobertRowneySr committed:
fix: wait for election to start

dobertRowneySr committed:
feat: set 1 minute session length for testing

dobertRowneySr committed:
fix: assertions

dobertRowneySr committed:
fix: formatting

dobertRowneySr committed:
fix: final fixes

- 30 seconds sessions
- wait for `era_blocks` for poa assertions
- wait for new era elections before Npos assertions

Commit:
A few POA->NPoS tests adjustments

Lezek123 reviewed (changes requested):
I fixed a few bugs in the tests and added some extra assertions, please take a look: https://github.com/ignazio-bovo/joystream/pull/8

Commit:
Merge pull request #8 from Lezek123/carthage-poa-npos-tests

A few POA->NPoS tests adjustments

Commit:
Merge branch 'carthage' into carthage-poa-npos-tests

Lezek123 commented on runtime/src/constants.rs:
```diff
@@ -37,7 +37,11 @@ pub const SECS_PER_BLOCK: Moment = MILLISECS_PER_BLOCK / 1000;
 
+#[cfg(feature = "testing_runtime")] // 30 seconds sessions for faster tests
+pub const EPOCH_DURATION_IN_BLOCKS: BlockNumber = MINUTES / 2;
+#[cfg(not(feature = "testing_runtime"))]
```
The feature has been renamed in https://github.com/Joystream/joystream/pull/4364, so this code no longer works
```suggestion
#[cfg(feature = "testing-runtime")] // 30 seconds sessions for faster tests
pub const EPOCH_DURATION_IN_BLOCKS: BlockNumber = MINUTES / 2;
#[cfg(not(feature = "testing-runtime"))]
```

Commit:
Update runtime/src/constants.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>
